helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java

```diff
@@ -241,13 +244,87 @@ private void refreshClusterConfig(final HelixDataAccessor accessor,
     if (_propertyDataChangedMap.get(HelixConstants.ChangeType.CLUSTER_CONFIG).getAndSet(false)) {
       _clusterConfig = accessor.getProperty(accessor.keyBuilder().clusterConfig());
       refreshedType.add(HelixConstants.ChangeType.CLUSTER_CONFIG);
+      // TODO: This is a temp function to clean up incompatible batched disabled instances format.
+      // Remove in later version.
+      if (checkBatchedDisabledInstanceFormat(_clusterConfig) && updateBatchDisableFormat(
+          accessor)) {
+        // read from zkz one more time
+        LogUtil.logInfo(logger, getClusterEventId(), String
+            .format("Clean ClusterConfig change for cluster %s, pipeline %s", _clusterName,
+                getPipelineName()));
+        _clusterConfig = accessor.getProperty(accessor.keyBuilder().clusterConfig());
+      }
       refreshAbnormalStateResolverMap(_clusterConfig);
     } else {
-      LogUtil.logInfo(logger, getClusterEventId(), String.format(
-          "No ClusterConfig change for cluster %s, pipeline %s", _clusterName, getPipelineName()));
+      LogUtil.logInfo(logger, getClusterEventId(), String
+          .format("No ClusterConfig change for cluster %s, pipeline %s", _clusterName,
+              getPipelineName()));
     }
   }
 
+  // TODO: This function is used to clean up incompatible batched disabled instances format for
+  // "DISABLED_INSTANCES" introduced in 1.0.3.0. This temp change shoul dbe reverted after 1.0.5.0 \
+  // or later version.
+  private boolean updateBatchDisableFormat(final HelixDataAccessor accessor) {
+    return accessor
+        .updateProperty(accessor.keyBuilder().clusterConfig(), new DataUpdater<ZNRecord>() {
+          @Override
+          public ZNRecord update(ZNRecord currentData) {
+            if (currentData == null) {
+              throw new HelixException(
+                  "Cluster: " + _clusterConfig.getClusterName() + ": cluster config is null");
+            }
```

Comment:
Similar to Junkai's comment above, let's try to combine the read-and-write pattern into one. If the currentData is null OR the format is good, we can just keep it. In such way we can combine the two methods into one updater.

Reply:
If we combine the 2 method into one updater, we will need to write either way. I would prefer have it separate.

Reply:
I think the problem with two separate steps is that, they are not transactional. Another concern is, the check is done on the local snapshot, but the modification is applied on the real zk data. 
Correct me if I understand wrong -- 
check(_clusterConfig) -> meanwhile clusterConfig is changed or removed -> updateBatchDisableFormat may fail or not needed

Reply:
This is correct, so we are checking the field again in updater

Reply:
Why not doing the check once in the updater, using the source of truth data?

Reply:
This will still issue a write request.

Reply:
That's true, we are not reducing the write, but that way we are having less read and completing the process in one step.

Reply:
Thanks for the suggestion, please correct me if I miss anything. 

If we do check and then write using update, then every time when we refresh cache for cluster config, we do one read to refresh, then check and finally read+write in updater if needed.
If we do check in updater, we will do one read to refresh and one read+write in updater no matter if we need to rewrite format. (we will write the cluster config back without any changes is most cases.)

Reply:
Thanks for the clarification. I had an impression that if updater returns the same value as old data, there will be no write. This might be a false assumption I made.
Let's keep it this way if a write is always triggered, but it's an interesting topic how we can improve property update to reduce write.